ICE `should have resolved vars before calling`

<!--
[31mICE[0m: Rustc ./D51E9A0748F2C6F08A9362336768DE5BCBDBAA7111422A99D2AC27F75786B86F.rs '-Zcrate-attr=feature(non_lifetime_binders) -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at /rustc-dev/518b428304e0008859cb1fd81d1ac20efb2a064a/compiler/rustc_trait_selection/src/traits/structural_normalize.rs:35:9: 'should have resolved vars before calling'', 'thread 'rustc' ($TID) panicked at /rustc-dev/518b428304e0008859cb1fd81d1ac20efb2a064a/compiler/rustc_trait_selection/src/traits/structural_normalize.rs:35:9: 'should have resolved vars before calling''
File: /tmp/im/D51E9A0748F2C6F08A9362336768DE5BCBDBAA7111422A99D2AC27F75786B86F.rs
-->

auto-reduced (treereduce-rust):
````rust
#![feature(non_lifetime_binders)]
pub trait T {
    fn t<Tail>(&self, _: F) {}
}

pub fn crash<V>(v: &V)
where
    for<F> F: T + 'static,
{
    v.t(|| {});
}

````

original:
````rust
pub trait T {
    fn t<Tail>(&self, _: F) {}
}

pub fn crash<V>(v: &V)
where
    for<F> F: T + 'static,
{
    v.t(|| {});
}

fn main() { }

````

Version information
````
rustc 1.93.0-nightly (518b42830 2025-11-16)
binary: rustc
commit-hash: 518b428304e0008859cb1fd81d1ac20efb2a064a
commit-date: 2025-11-16
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5

````



Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zcrate-attr=feature(non_lifetime_binders)`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0412]: cannot find type `F` in this scope
  --> /tmp/icemaker_global_tempdir.60fr4bItXUXX/rustc_testrunner_tmpdir_reporting.8KrpuXfoHl78/mvce.rs:2:26
   |
 2 |     fn t<Tail>(&self, _: F) {}
   |                          ^ help: a trait with a similar name exists: `Fn`
   |
  ::: /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/ops/function.rs:76:1
   |
76 | pub const trait Fn<Args: Tuple>: FnMut<Args> {
   | -------------------------------------------- similarly named trait `Fn` defined here

warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
 --> <crate attribute>:1:12
  |
1 | #![feature(non_lifetime_binders)]
  |            ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
  = note: `#[warn(incomplete_features)]` on by default

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.60fr4bItXUXX/rustc_testrunner_tmpdir_reporting.8KrpuXfoHl78/mvce.rs:10:2
   |
10 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.60fr4bItXUXX/rustc_testrunner_tmpdir_reporting.8KrpuXfoHl78/mvce.rs`


thread 'rustc' (3684634) panicked at /rustc-dev/518b428304e0008859cb1fd81d1ac20efb2a064a/compiler/rustc_trait_selection/src/traits/structural_normalize.rs:35:9:
should have resolved vars before calling
stack backtrace:
   0:     0x7f117e535b33 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h5ae4ae86875c00e9
   1:     0x7f117ec01d9c - core::fmt::write::he58289da8ad88ac4
   2:     0x7f117e4eb0b6 - std::io::Write::write_fmt::ha47e198308c7c52d
   3:     0x7f117e50233c - std::panicking::default_hook::{{closure}}::h6b3da92cff9dfda6
   4:     0x7f117e501e93 - std::panicking::default_hook::hc2cfe13e40265481
   5:     0x7f117d53fbf7 - std[d45af2a1259e8800]::panicking::update_hook::<alloc[58b5be5fbc54de9e]::boxed::Box<rustc_driver_impl[d2f42e81b9c37b75]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f117e5026f2 - std::panicking::panic_with_hook::h2172996bacdba2d7
   7:     0x7f117e5024ba - std::panicking::panic_handler::{{closure}}::hc36cea0dc92c48aa
   8:     0x7f117e4fc3f9 - std::sys::backtrace::__rust_end_short_backtrace::hd4b38282c54e463b
   9:     0x7f117e4dd6dd - __rustc[f20bc1530c389e91]::rust_begin_unwind
  10:     0x7f117b49c23c - core::panicking::panic_fmt::hda6de8e6488d2e11
  11:     0x7f117f5c72db - <rustc_infer[6f4e77a6db3e3b7c]::infer::InferCtxt>::probe::<rustc_hir_typeck[8926b67bfda1115c]::method::probe::ProbeResult, <rustc_hir_typeck[8926b67bfda1115c]::method::probe::ProbeContext>::consider_probe::{closure#0}::{closure#0}>
  12:     0x7f117f5b5b61 - <rustc_hir_typeck[8926b67bfda1115c]::method::probe::ProbeContext>::pick_all_method
  13:     0x7f117fce5f86 - <rustc_hir_typeck[8926b67bfda1115c]::fn_ctxt::FnCtxt>::lookup_probe
  14:     0x7f117ee215f0 - <rustc_hir_typeck[8926b67bfda1115c]::fn_ctxt::FnCtxt>::check_expr_method_call
  15:     0x7f117ee5538f - <rustc_hir_typeck[8926b67bfda1115c]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16:     0x7f117ee4cb64 - <rustc_hir_typeck[8926b67bfda1115c]::fn_ctxt::FnCtxt>::check_expr_block
  17:     0x7f117ee553ab - <rustc_hir_typeck[8926b67bfda1115c]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  18:     0x7f117ee17a49 - rustc_hir_typeck[8926b67bfda1115c]::check::check_fn
  19:     0x7f117f83abc6 - rustc_hir_typeck[8926b67bfda1115c]::typeck_with_inspect::{closure#0}
  20:     0x7f117f8395c0 - rustc_query_impl[88cece6c3b652f81]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[88cece6c3b652f81]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[65c3dfc665d99b31]::query::erase::Erased<[u8; 8usize]>>
  21:     0x7f117edf41bc - rustc_query_system[3302b3c6b75e0c59]::query::plumbing::try_execute_query::<rustc_query_impl[88cece6c3b652f81]::DynamicConfig<rustc_data_structures[7b18bb195bd0e66b]::vec_cache::VecCache<rustc_span[55bb1a521e26a22e]::def_id::LocalDefId, rustc_middle[65c3dfc665d99b31]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[3302b3c6b75e0c59]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[88cece6c3b652f81]::plumbing::QueryCtxt, false>
  22:     0x7f117edf3b49 - rustc_query_impl[88cece6c3b652f81]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  23:     0x7f117f70530b - <rustc_middle[65c3dfc665d99b31]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[f6e852dd36e2adf2]::check_crate::{closure#2}>::{closure#0}
  24:     0x7f117f70463e - rustc_hir_analysis[f6e852dd36e2adf2]::check_crate
  25:     0x7f117ededab7 - rustc_interface[67c0c783642d6c76]::passes::analysis
  26:     0x7f117eded775 - rustc_query_impl[88cece6c3b652f81]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[88cece6c3b652f81]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[65c3dfc665d99b31]::query::erase::Erased<[u8; 0usize]>>
  27:     0x7f117fce2a94 - rustc_query_system[3302b3c6b75e0c59]::query::plumbing::try_execute_query::<rustc_query_impl[88cece6c3b652f81]::DynamicConfig<rustc_query_system[3302b3c6b75e0c59]::query::caches::SingleCache<rustc_middle[65c3dfc665d99b31]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[88cece6c3b652f81]::plumbing::QueryCtxt, false>
  28:     0x7f117fce26ca - rustc_query_impl[88cece6c3b652f81]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7f117ff3eb90 - <rustc_interface[67c0c783642d6c76]::passes::create_and_enter_global_ctxt<core[7c9280adad8818af]::option::Option<rustc_interface[67c0c783642d6c76]::queries::Linker>, rustc_driver_impl[d2f42e81b9c37b75]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[7c9280adad8818af]::ops::function::FnOnce<(&rustc_session[54c87533a8d99824]::session::Session, rustc_middle[65c3dfc665d99b31]::ty::context::CurrentGcx, alloc[58b5be5fbc54de9e]::sync::Arc<rustc_data_structures[7b18bb195bd0e66b]::jobserver::Proxy>, &std[d45af2a1259e8800]::sync::once_lock::OnceLock<rustc_middle[65c3dfc665d99b31]::ty::context::GlobalCtxt>, &rustc_data_structures[7b18bb195bd0e66b]::sync::worker_local::WorkerLocal<rustc_middle[65c3dfc665d99b31]::arena::Arena>, &rustc_data_structures[7b18bb195bd0e66b]::sync::worker_local::WorkerLocal<rustc_hir[afdf7e16091c21f7]::Arena>, rustc_driver_impl[d2f42e81b9c37b75]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  30:     0x7f117fdd6105 - rustc_interface[67c0c783642d6c76]::interface::run_compiler::<(), rustc_driver_impl[d2f42e81b9c37b75]::run_compiler::{closure#0}>::{closure#1}
  31:     0x7f117fd2afba - std[d45af2a1259e8800]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[67c0c783642d6c76]::util::run_in_thread_with_globals<rustc_interface[67c0c783642d6c76]::util::run_in_thread_pool_with_globals<rustc_interface[67c0c783642d6c76]::interface::run_compiler<(), rustc_driver_impl[d2f42e81b9c37b75]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  32:     0x7f117fd2aca8 - <<std[d45af2a1259e8800]::thread::Builder>::spawn_unchecked_<rustc_interface[67c0c783642d6c76]::util::run_in_thread_with_globals<rustc_interface[67c0c783642d6c76]::util::run_in_thread_pool_with_globals<rustc_interface[67c0c783642d6c76]::interface::run_compiler<(), rustc_driver_impl[d2f42e81b9c37b75]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[7c9280adad8818af]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  33:     0x7f117fd309af - std::sys::thread::unix::Thread::new::thread_start::h544843a2b008d4a8
  34:     0x7f11796969cb - <unknown>
  35:     0x7f117971aa0c - <unknown>
  36:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.93.0-nightly (518b42830 2025-11-16) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z crate-attr=feature(non_lifetime_binders) -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck] type-checking `crash`
#1 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 2 previous errors; 1 warning emitted

Some errors have detailed explanations: E0412, E0601.
For more information about an error, try `rustc --explain E0412`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `crash`
#1 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-non_lifetime_binders

--- a/tests/ui/traits/late-bound-type-method-probe.rs
+++ b/tests/ui/traits/late-bound-type-method-probe.rs
@@ -1,0 +1,19 @@
+// A method probe involving a late-bound type parameter should report normal
+// diagnostics instead of ICEing while structurally normalizing obligations.
+
+#![feature(non_lifetime_binders)]
+#![allow(incomplete_features)]
+
+pub trait T {
+    fn t<Tail>(&self, _: F) {}
+    //~^ ERROR cannot find type `F` in this scope
+}
+
+pub fn crash<V>(v: &V)
+where
+    for<F> F: T + 'static,
+{
+    v.t(|| {});
+}
+
+fn main() {}

--- a/tests/ui/traits/late-bound-type-method-probe.stderr
+++ b/tests/ui/traits/late-bound-type-method-probe.stderr
@@ -1,0 +1,17 @@
+error[E0425]: cannot find type `F` in this scope
+  --> $DIR/late-bound-type-method-probe.rs:8:26
+   |
+LL |     fn t<Tail>(&self, _: F) {}
+   |                          ^
+   |
+  --> $SRC_DIR/core/src/ops/function.rs:LL:COL
+   |
+   = note: similarly named trait `Fn` defined here
+help: a trait with a similar name exists
+   |
+LL |     fn t<Tail>(&self, _: Fn) {}
+   |                           +
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0425`.

